### DISCUSSION TITLE
Data importer behind auth-proxy failed to receive data from firefly-iii

### DISCUSSION THREAD

**Discussion by shizunge** [firefly-iii/firefly-iii#5787] [Firefly III] [answered]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)
- [X] My request is not listed as [a very good idea, but unfortunately...](https://docs.firefly-iii.org/firefly-iii/about-firefly-iii/what-its-not/)
- [X] I've used [the search](https://github.com/firefly-iii/firefly-iii/issues?q=is%3Aissue) and this has not been discussed before.

### Description

I am "upgrading" csv importer to data importer v0.8.0, and kept get connection error while authorization. There is no enough error log message to help me debug.

Eventually I found the cause: there was a callback from firefly-iii to the data importer, which was new in the data importer. I had an auth-proxy in front of my data importer, and the traffic from firefly-iii to the data importer did not go through the docker network, where both containers connected to, instead the traffic went through the public network and failed the authorization. As a result, data importer never heard back from firefly-iii.

### Solution

1.   Improve the error message of data importer, when not hearing back from firefly-iii.
2.   If the callback is necessary, could you please add a configure let user select the base address of the callback? Especifically we have been able to specify the firefly-iii URL via FIREFLY_III_URL.  I am accessing the data importer via a public address e.g. https://data-importer.example.com but I want the firefly-iii and data-importer communicate via the docker/local network, e.g. http://data-importer:8080 & http://firefly-iii:8080
3. If we cannot do 2, the best way is actually letting firefly-iii provide authorization information. It would help even if it works only for the http basic authorization. But I think it may be much more work than 2.

### What are alternatives?

Comparing to csv-importer, this is new. Alternative is to not expect some data from firefly-iii like the csv-importer.

**Comment by JC5** [comment]
1. There is a timeout message in the logs, but all I see in #5772 is Apache-related logging. What are your `.env` settings related to logging?
2. This is already possible. Please read the `.env.example` file.

Converted to an issue because I do not believe this is a bug.

**Reply by shizunge** [reply]
My .env has only the time zone info `TZ=America/Los_Angeles`

Maybe I did not explain myself clearly. There is not problem with `FIREFLY_III_URL`.
The problem is about something like `DATA_IMPORTER_URL` missing.
I did not find anything useful in the `.env.example`. Maybe `ASSET_URL`? But it is not about the base URL and massed up the UI. `APP_URL` is said unused and had no effect in my experiments.

Since I setup via docker, I can actually access data-importer via 2 URLs.
1. http://172.31.240.13:8080
    This one is the containers IP address. And I can use this IP only on the machine where the container is running. No auth-proxy.
2. https://data-importer-public-address.com
    This one can be accessed anywhere, but there is authorization proxy in front of the services.

I run the following experiments.
1. Access data-importer via http://172.31.240.13:8080 and data-importer authorized.
Log is
```
today at 21:38:50[2022-02-13 21:38:50] production.DEBUG: Now in App\Http\Controllers\IndexController::index  
today at 21:38:50[2022-02-13 21:38:50] production.DEBUG: Now in App\Http\Controllers\IndexController::index  
today at 21:38:50[2022-02-13 21:38:50] production.DEBUG: App\Services\Shared\Authentication\SecretManager::hasValidSecrets  
today at 21:38:50[2022-02-13 21:38:50] production.DEBUG: App\Services\Shared\Authentication\SecretManager::hasValidSecrets  
today at 21:38:50[2022-02-13 21:38:50] production.DEBUG: No valid secrets, redirect to token.index  
today at 21:38:50[2022-02-13 21:38:50] production.DEBUG: No valid secrets, redirect to token.index  
today at 21:38:51 172.31.240.1 - - [13/Feb/2022:21:38:45 -0800] "GET / HTTP/1.1" 302 1535 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:96.0) Gecko/20100101 Firefox/96.0"
today at 21:38:51 172.31.240.1 - - [13/Feb/2022:21:38:45 -0800] "GET / HTTP/1.1" 302 1535 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:96.0) Gecko/20100101 Firefox/96.0"
today at 21:38:56[2022-02-13 21:38:56] production.DEBUG: Now at App\Http\Controllers\TokenController::index  
today at 21:38:57[2022-02-13 21:38:57] production.DEBUG: No access token in hasAccessToken(), will return config variable.  
today at 21:38:57[2022-02-13 21:38:57] production.DEBUG: No client id in hasClientId(), will return config variable.  
today at 21:38:57[2022-02-13 21:38:57] production.DEBUG: No base url in getBaseUrl(), will return config variable.  
today at 21:38:57[2022-02-13 21:38:57] production.DEBUG: No vanity url in getVanityUrl(), will return config variable.  
today at 21:38:57[2022-02-13 21:38:57] production.INFO: The following configuration information was found:  
today at 21:38:57[2022-02-13 21:38:57] production.INFO: Personal Access Token: "eyJ0eXAiOiJKV1QiLCJhbGciO" (limited to 25 chars if present)  
today at 21:38:57[2022-02-13 21:38:57] production.INFO: Client ID            : "0"  
today at 21:38:57[2022-02-13 21:38:57] production.INFO: Base URL             : "http://firefly-iii:8080"  
today at 21:38:57[2022-02-13 21:38:57] production.INFO: Vanity URL           : "https://${REAL_ADDRESS_REPLACED}"  
today at 21:38:57[2022-02-13 21:38:57] production.DEBUG: Found personal access token + URL "http://firefly-iii:8080" in config, set cookie and return to index.  
today at 21:38:57 172.31.240.1 - - [13/Feb/2022:21:38:54 -0800] "GET /token HTTP/1.1" 302 4641 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:96.0) Gecko/20100101 Firefox/96.0"
today at 21:38:58[2022-02-13 21:38:58] production.DEBUG: Now in App\Http\Controllers\IndexController::index  
today at 21:38:58[2022-02-13 21:38:58] production.DEBUG: App\Services\Shared\Authentication\SecretManager::hasValidSecrets 
today at 21:38:58 172.31.240.1 - - [13/Feb/2022:21:38:58 -0800] "GET / HTTP/1.1" 200 3219 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:96.0) Gecko/20100101 Firefox/96.0"
today at 21:39:03 172.31.240.1 - - [13/Feb/2022:21:39:02 -0800] "GET /js/bootstrap.min.js?version=0.8.0 HTTP/1.1" 200 16729 "http://172.31.240.13:8080/" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:96.0) Gecko/20100101 Firefox/96.0"
today at 21:39:03 172.31.240.1 - - [13/Feb/2022:21:39:02 -0800] "GET /fa/css/all.min.css HTTP/1.1" 200 13028 "http://172.31.240.13:8080/" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:96.0) Gecko/20100101 Firefox/96.0"
today at 21:39:03 172.31.240.1 - - [13/Feb/2022:21:39:02 -0800] "GET /js/jquery-3.6.0.min.js?version=0.8.0 HTTP/1.1" 200 31271 "http://172.31.240.13:8080/" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:96.0) Gecko/20100101 Firefox/96.0"
today at 21:39:03 172.31.240.1 - - [13/Feb/2022:21:39:02 -0800] "GET /css/bootstrap-dark.min.css?version=0.8.0 HTTP/1.1" 200 31464 "http://172.31.240.13:8080/" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:96.0) Gecko/20100101 Firefox/96.0"
today at 21:39:03 172.31.240.1 - - [13/Feb/2022:21:39:02 -0800] "GET /js/popper.min.js?version=0.8.0 HTTP/1.1" 200 7080 "http://172.31.240.13:8080/" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:96.0) Gecko/20100101 Firefox/96.0"
today at 21:39:03 172.31.240.1 - - [13/Feb/2022:21:39:03 -0800] "GET /fa/webfonts/fa-solid-900.woff2 HTTP/1.1" 200 79748 "http://172.31.240.13:8080/fa/css/all.min.css" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:96.0) Gecko/20100101 Firefox/96.0"
today at 21:39:06[2022-02-13 21:39:06] production.DEBUG: Now at App\Http\Controllers\TokenController::doValidate  
today at 21:39:06[2022-02-13 21:39:06] production.DEBUG: isReadyForStep returns true because $flow is null  
today at 21:39:06[2022-02-13 21:39:06] production.DEBUG: isReadyForStep returns true because $flow is null  
today at 21:39:06[2022-02-13 21:39:06] production.DEBUG: Now at App\Services\Spectre\AuthenticationValidator::validate  
today at 21:39:06[2022-02-13 21:39:06] production.DEBUG: No Spectre App ID in hasAppId(), will return config variable.  
today at 21:39:06[2022-02-13 21:39:06] production.DEBUG: No Spectre secret in hasSecret(), will return config variable.  
today at 21:39:06[2022-02-13 21:39:06] production.DEBUG: Now at App\Services\Nordigen\AuthenticationValidator::validate  
today at 21:39:06[2022-02-13 21:39:06] production.DEBUG: No Nordigen ID in hasId(), will return config variable.  
today at 21:39:06[2022-02-13 21:39:06] production.DEBUG: No Nordigen key in hasKey(), will return config variable.  
today at 21:39:06 172.31.240.1 - - [13/Feb/2022:21:39:03 -0800] "GET /validate/spectre HTTP/1.1" 200 1128 "http://172.31.240.13:8080/" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:96.0) Gecko/20100101 Firefox/96.0"
today at 21:39:06 172.31.240.1 - - [13/Feb/2022:21:39:03 -0800] "GET /validate/nordigen HTTP/1.1" 200 1128 "http://172.31.240.13:8080/" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:96.0) Gecko/20100101 Firefox/96.0"
today at 21:39:17[2022-02-13 21:39:17] production.DEBUG: Result is {"result":"OK","message":null} 
today at 21:39:17 172.31.240.1 - - [13/Feb/2022:21:39:03 -0800] "GET /token/validate HTTP/1.1" 200 1139 "http://172.31.240.13:8080/" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:96.0) Gecko/20100101 Firefox/96.0"
today at 21:40:55 172.31.240.1 - - [13/Feb/2022:21:40:55 -0800] "GET /favicon.ico HTTP/1.1" 200 280 "http://172.31.240.13:8080/" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:96.0) Gecko/20100101 Firefox/96.0
```

Some one (I think it was firefly-iii) was trying to access
* http://172.31.240.13:8080/validate/spectre
* http://172.31.240.13:8080/validate/nordigen
* http://172.31.240.13:8080/token/validate

As it was an IP address in the docker network, there was no auth-proxy in the path, and these accesses succeed.

2. Access via https://data-importer-public-address.com
And run "Re-authorization". No error message, and auth failed. Log is
```
today at 21:42:03[2022-02-13 21:42:03] production.DEBUG: Now in App\Http\Controllers\IndexController::index  
today at 21:42:03[2022-02-13 21:42:03] production.DEBUG: App\Services\Shared\Authentication\SecretManager::hasValidSecrets  
today at 21:42:03 172.31.240.47 - - [13/Feb/2022:21:42:02 -0800] "GET / HTTP/1.1" 200 3159 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:97.0) Gecko/20100101 Firefox/97.0"
today at 21:42:18[2022-02-13 21:42:18] production.DEBUG: Now at App\Http\Controllers\IndexController::reset  
today at 21:42:19 172.31.240.47 - - [13/Feb/2022:21:42:18 -0800] "GET /reset HTTP/1.1" 302 2662 "https://${REAL_ADDRESS_REPLACED}/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:97.0) Gecko/20100101 Firefox/97.0"
today at 21:42:19[2022-02-13 21:42:19] production.DEBUG: Now in App\Http\Controllers\IndexController::index  
today at 21:42:19[2022-02-13 21:42:19] production.DEBUG: App\Services\Shared\Authentication\SecretManager::hasValidSecrets  
today at 21:42:19[2022-02-13 21:42:19] production.DEBUG: No valid secrets, redirect to token.index  
today at 21:42:19 172.31.240.47 - - [13/Feb/2022:21:42:19 -0800] "GET / HTTP/1.1" 302 1509 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:97.0) Gecko/20100101 Firefox/97.0"
today at 21:42:19[2022-02-13 21:42:19] production.DEBUG: Now at App\Http\Controllers\TokenController::index  
today at 21:42:19[2022-02-13 21:42:19] production.DEBUG: No access token in hasAccessToken(), will return config variable.  
today at 21:42:19[2022-02-13 21:42:19] production.DEBUG: No client id in hasClientId(), will return config variable.  
today at 21:42:19[2022-02-13 21:42:19] production.DEBUG: No base url in getBaseUrl(), will return config variable.  
today at 21:42:19[2022-02-13 21:42:19] production.INFO: The following configuration information was found:  
today at 21:42:19[2022-02-13 21:42:19] production.INFO: Personal Access Token: "eyJ0eXAiOiJKV1QiLCJhbGciO" (limited to 25 chars if present)  
today at 21:42:19[2022-02-13 21:42:19] production.INFO: Client ID            : "0"  
today at 21:42:19[2022-02-13 21:42:19] production.INFO: Base URL             : "http://firefly-iii:8080"  
today at 21:42:19[2022-02-13 21:42:19] production.INFO: Vanity URL           : "http://firefly-iii:8080"  
today at 21:42:19[2022-02-13 21:42:19] production.DEBUG: Found personal access token + URL "http://firefly-iii:8080" in config, set cookie and return to index.  
today at 21:42:19 172.31.240.47 - - [13/Feb/2022:21:42:19 -0800] "GET /token HTTP/1.1" 302 4616 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:97.0) Gecko/20100101 Firefox/97.0"
today at 21:42:20[2022-02-13 21:42:20] production.DEBUG: Now in App\Http\Controllers\IndexController::index  
today at 21:42:20[2022-02-13 21:42:20] production.DEBUG: App\Services\Shared\Authentication\SecretManager::hasValidSecrets  
today at 21:42:20 172.31.240.47 - - [13/Feb/2022:21:42:20 -0800] "GET / HTTP/1.1" 200 3159 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:97.0) Gecko/20100101 Firefox/97.0"
```

With help from my auth-proxy log, some one (I think it was firefly-iii) was trying to access
* https://data-importer-public-address.com/validate/spectre
* https://data-importer-public-address.com/validate/nordigen
* https://data-importer-public-address.com/token/validate

But failed authorization. My auth-proxy returned 401.


It seems the base address of `/token/validate` `/validate/nordigen` and `/validate/spectre` depended on which address I used to access data-importer.

My question is:  Could we make this base address a fixed value?

Also note, you are talking about "There is a timeout message". I think you meant the timeout to reach `FIREFLY_III_URL`? There is no problem for data-importer to reach `FIREFLY_III_URL` and the auth error appear about 1-2 seconds after I clicked the "Re-authorization" button, much shorter than the `CONNECTION_TIMEOUT` I found in the `.env.example`. The problem is we got 401 to reach `https://data-importer-public-address.com/token/validate`. No error message in firefly-iii, no error message in data-importer.

**Reply by shizunge** [reply]
Narrow down the access to `https://data-importer-public-address.com/token/validate` to the frontend. i.e. the codes running in the browser.

After I added `TRUSTED_PROXIES=**`, it worked.

But I do not understand why my auth-proxy reported 401. It looked like the browser did not send cookies to the auth-proxy.

Also could we improve the front end side error message? I have to open the developer tool to see the error message, I do not think it is a good way to read error message for a normal user.

**Comment by shizunge** [comment, marked as answer]
Answer to myself.

After I added `TRUSTED_PROXIES=**`, it worked.

The front end (codes in the browser) got http 401 error, and did not display it. So no error message on the server side.

**Reply by JC5** [reply]
Thanks! I'll do some debugging to see if I can catch all of the http error codes that can be thrown at the importer.
